src/threadpool.c

```diff
@@ -310,3 +328,52 @@ int uv_cancel(uv_req_t* req) {
 
   return uv__work_cancel(loop, req, wreq);
 }
+
+
+static void report(enum estage stage, int lock) {
+  QUEUE* q;
+  uv_queue_stats_t* s;
+  int length;
+  int threads;
+
+  if (lock)
+    uv_mutex_lock(&mutex);
+
+  if (!QUEUE_EMPTY(&stats)) {
+    length = wq_length;
+    threads = idle_threads;
+
+    QUEUE_FOREACH(q, &stats) {
+      s = QUEUE_DATA(q, struct uv_queue_stats_s, q);
+      switch (stage) {
+        case SUBMIT:
+          s->submit_cb(length, threads, s->data);
+          break;
+        case START:
+          s->start_cb(length, threads, s->data);
+          break;
+        case DONE:
+          s->done_cb(length, threads, s->data);
+          break;
+        default: abort();
+      }
+    }
+  }
+
+  if (lock)
+    uv_mutex_unlock(&mutex);
+}
```

Comment:
This function still invokes the callbacks with the lock held.

Also, programmer economics: you can't call `uv_queue_stats_stop()` from a callback now (although for that to work properly, you can't use `QUEUE_FOREACH()`.)